### PR TITLE
Adding driver for four-wire resistive touchscreen

### PR DESCRIPTION
**Pull request by bgould**
This is a driver for a simple resistive four-wire touchscreen like the one found on PyPortal and many inexpensive displays.

The code works but this PR is still WIP as I am doing some final testing I still need to:

Update README
Add example(s)
Validate the logic

Also I would like to get some feedback on the package layout and interface.  It might be premature for a "standard" interface but I added on called "Pointer" in the touch package.

### CODE REVIEW

**Comment by bgould** [comment]
Added smoke test and removed WIP from title.  I think this is ready for review please.

File: Makefile [hunk 1]
```diff
@@ -85,6 +85,8 @@ smoke-test:
 	@md5sum ./build/test.hex
 	tinygo build -size short -o ./build/test.hex -target=circuitplay-express ./examples/thermistor/main.go
 	@md5sum ./build/test.hex
+	tinygo build -size short -o ./build/test.hex -target=pyportal ./examples/touch/resistive/fourwire/main.go
+	@md5sum ./build/test.hex
```
Comment (aykevl) [inline Makefile]: The PyPortal example should also be included here in the smoke tests.

File: examples/touch/resistive/pyportal_touchpaint/main.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+	}
+
+	if (int16(touch.Y) - penRadius) > boxSize {
+		display.FillRectangle(
+			int16(touch.X), int16(touch.Y), penRadius*2, penRadius*2, currentColor)
+	}
+}
+
+/**** End of Touchpaint demo ************************************************/
```
Comment (aykevl) [inline examples/touch/resistive/pyportal_touchpaint/main.go]: this looks a bit out of place in Go source code

File: touch/resistive/fourwire.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+func (res *FourWireTouchscreen) GetTouchPoint() (p touch.Point) {
+	p.X = int(res.ReadX())
+	p.Y = int(res.ReadY())
+	p.Z = int(res.ReadZ())
+	return
+}
+
+func (res *FourWireTouchscreen) ReadX() uint16 {
+	res.YM.Configure(machine.PinConfig{machine.PinInput})
+	res.YM.Low()
```
Comment (aykevl) [inline touch/resistive/fourwire.go]: You cannot normally set a pin high or low when it's an input. What is the intention here?

File: touch/resistive/fourwire.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+type FourWireTouchscreen struct {
+	YP machine.Pin
+	YM machine.Pin
+	XP machine.Pin
+	XM machine.Pin
+
+	yp machine.ADC
+	xm machine.ADC
+	xp machine.ADC
```
Comment (aykevl) [inline touch/resistive/fourwire.go]: It looks like these are essentially duplicates of the pins declared above. It should be possible (and more efficient) to declare them inline:

```
machine.ADC{res.XP}.Get()
```

File: examples/touch/resistive/pyportal_touchpaint/main.go (hunk 2 above)
Comment (bgould) [reply examples/touch/resistive/pyportal_touchpaint/main.go]: Oh yes this definitely doesn't belong there, I will remove it.

File: touch/resistive/fourwire.go (hunk 4 above)
Comment (bgould) [reply touch/resistive/fourwire.go]: Ok yes, I think that would be clearer.  I wasn't sure if creating a new instance of the struct for each call to Get() would have a negative impact on anything, so I went with just keeping a copy of it around.  I will change it over to the way you've suggested

File: Makefile (hunk 1 above)
Comment (bgould) [reply Makefile]: Will do

**Comment by bgould** [comment]
Thank you very much for the feedback @aykevl , I think these are all good suggestions.  

> * Externally visible names must have documentation.

Yes, no problem, I will add.

> * What's the scale used here? It _appears_ be be 0..1023, but why this particular scale? For future proofing, it may be better to store the value in the range 0..65535 (the full uint16 range), similar to the return value of the ADC. Or otherwise some human-understandable range such as 0..1000 or 0..10000 (perhaps returning the result as an int/int32 instead of uint16).

This makes sense; I think I was thinking that scaling it down to 10 bits would be the lowest common denominator for AVR/Arduino or something, but what you're saying makes sense and wouldn't force lossiness on hardware that supports better precision.

> * In other drivers, I've prefixed sensor reading methods with `Read`, I think it's best to do that here too (as `Get` suggests a getter), such as `ReadTouchPoint`.

Sounds good to me, I will adjust the code accordingly

> Also, what's the reason for exporting the `ReadX` etc methods? It seems to me that all you need is a single method to read the touch point.

Originally I think I was exposing those to provide a low-level way to read the individual values, and was going to have GetTouchPoint() possibly implement things like debouncing and converting everything to a common scale.  I practice as I moved code around, things obviously didn't turn out like that so you're right it doesn't make a ton of sense to export these.

I will give this another shot and clean things up based on your feedback.

File: Makefile (hunk 1 above)
Comment (bgould) [reply Makefile]: This has been addressed in the latest set of commits.

File: touch/resistive/fourwire.go (hunk 4 above)
Comment (bgould) [reply touch/resistive/fourwire.go]: This is addressed in the latest set of commits.

File: touch/resistive/fourwire.go (hunk 3 above)
Comment (bgould) [reply touch/resistive/fourwire.go]: This is addressed in the latest set of commits.

File: examples/touch/resistive/pyportal_touchpaint/main.go (hunk 2 above)
Comment (bgould) [reply examples/touch/resistive/pyportal_touchpaint/main.go]: This is addressed in the latest set of commits.

**Comment by deadprogram** [comment]
Hi @bgould seems like you left in the `ReadX()` methods. Are those supposed to be removed or is it fine the way it is? I do not have a strong opinion either way.

**Comment by bgould** [comment]
> Hi @bgould seems like you left in the `ReadX()` methods. Are those supposed to be removed or is it fine the way it is? I do not have a strong opinion either way.

I ended up leaving them in but changed their purpose slightly and documented it accordingly.  Before the values returned from Read* were equivalent as if you called ReadTouchPoint which returns all three.  Now, ReadTouchPoint() returns these values but averages them across several reads.  The Read* methods on the other return the "raw" value from a single read.  I left these as exported in case someone wants to embed the struct and override ReadTouchPoint.

ReadTouchPoint is the method that defines the touch.Pointer interface and really the only one that most client code should care about.

**Comment by deadprogram** [comment]
Finally got my PyPortal Pynt and had a chance to try this out. Worked great, the mini-paint application is too cool!

Now squashing/merging, thanks for another very nice contribution @bgould

**Pull request merged**